File: problems/tier2-intermediate/015-bookmyshow/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <set>
#include <algorithm>
using namespace std;

// ─── Data Structures ────────────────────────────────────────────────────────

enum class SeatStatus { AVAILABLE, LOCKED, BOOKED };

struct Seat {
    int row;
    int col;
    SeatStatus status;
    string lockedBy;
    long lockExpiry;
    string bookedBy;
};

struct Show {
    string id;
    string theaterId;
    string movie;
    string time;
    int rows;
    int cols;
    vector<vector<Seat>> seats;
};

struct Theater {
    string id;
    string name;
    string city;
};

struct Booking {
    string id;
    string showId;
    string userId;
    vector<pair<int,int>> seatPositions;
};

struct SeatLock {
    string id;
    string showId;
    string userId;
    vector<pair<int,int>> seatPositions;
    long expiry;
    bool confirmed;
    bool released;
};

// ─── Booking System ────────────────────────────────────────────────────────

class BookingSystem {
    unordered_map<string, Theater> theaters;
    unordered_map<string, Show> shows;
    unordered_map<string, Booking> bookings;
    unordered_map<string, SeatLock> locks;
    unordered_map<string, set<string>> cityMovies;
    int lockCounter = 0;

    void expireSeat(Seat& seat, long currentTime) {
        if (seat.status == SeatStatus::LOCKED && currentTime >= seat.lockExpiry) {
            seat.status = SeatStatus::AVAILABLE;
            seat.lockedBy = "";
            seat.lockExpiry = 0;
        }
    }

    bool isSeatAvailable(const Seat& seat, long currentTime) const {
        if (seat.status == SeatStatus::AVAILABLE) return true;
        if (seat.status == SeatStatus::LOCKED && currentTime >= seat.lockExpiry)
            return true;
        return false;
    }

public:
    void addTheater(const string& theaterId, const string& name,
                    const string& city) {
        theaters[theaterId] = {theaterId, name, city};
    }

    void addShow(const string& showId, const string& theaterId,
                 const string& movie, const string& time, int rows, int cols) {
        Show show{showId, theaterId, movie, time, rows, cols, {}};
        show.seats.resize(rows, vector<Seat>(cols));
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                show.seats[r][c] = {r, c, SeatStatus::AVAILABLE, "", 0, ""};
        shows[showId] = show;
        if (theaters.count(theaterId))
            cityMovies[theaters[theaterId].city].insert(movie);
    }

    vector<string> searchMovies(const string& city) {
        if (!cityMovies.count(city)) return {};
        return vector<string>(cityMovies[city].begin(), cityMovies[city].end());
    }

    vector<pair<int,int>> getAvailableSeats(const string& showId,
                                            long currentTime = 0) {
        if (!shows.count(showId)) return {};
        auto& show = shows[showId];
        vector<pair<int,int>> result;
        for (int r = 0; r < show.rows; r++)
            for (int c = 0; c < show.cols; c++) {
                expireSeat(show.seats[r][c], currentTime);
                if (show.seats[r][c].status == SeatStatus::AVAILABLE)
                    result.push_back({r, c});
            }
        return result;
    }

    bool bookSeats(const string& bookingId, const string& showId,
                   const vector<pair<int,int>>& seatPositions,
                   const string& userId, long currentTime = 0) {
        if (!shows.count(showId)) return false;
        auto& show = shows[showId];
        // Atomic check: all seats must be available
        for (auto& [r, c] : seatPositions) {
            if (r < 0 || r >= show.rows || c < 0 || c >= show.cols)
                return false;
            expireSeat(show.seats[r][c], currentTime);
            if (show.seats[r][c].status != SeatStatus::AVAILABLE)
                return false;
        }
        // Book all seats
        for (auto& [r, c] : seatPositions) {
            show.seats[r][c].status = SeatStatus::BOOKED;
            show.seats[r][c].bookedBy = userId;
        }
        bookings[bookingId] = {bookingId, showId, userId, seatPositions};
        return true;
    }

    string lockSeats(const string& showId,
                     const vector<pair<int,int>>& seatPositions,
                     const string& userId, int ttlMinutes, long currentTime) {
        if (!shows.count(showId)) return "";
        auto& show = shows[showId];
        // Check all seats are available
        for (auto& [r, c] : seatPositions) {
            if (r < 0 || r >= show.rows || c < 0 || c >= show.cols)
                return "";
            expireSeat(show.seats[r][c], currentTime);
            if (show.seats[r][c].status != SeatStatus::AVAILABLE)
                return "";
        }
        string lockId = "LOCK_" + to_string(++lockCounter);
        long expiry = currentTime + ttlMinutes * 60;
        // Lock all seats
        for (auto& [r, c] : seatPositions) {
            show.seats[r][c].status = SeatStatus::LOCKED;
            show.seats[r][c].lockedBy = userId;
            show.seats[r][c].lockExpiry = expiry;
        }
        locks[lockId] = {lockId, showId, userId, seatPositions,
                         expiry, false, false};
        return lockId;
    }

    bool confirmBooking(const string& lockId, long currentTime) {
        if (!locks.count(lockId)) return false;
        auto& lock = locks[lockId];
        if (lock.confirmed || lock.released) return false;
        if (currentTime >= lock.expiry) {
            // Lock expired — release seats
            if (shows.count(lock.showId)) {
                auto& show = shows[lock.showId];
                for (auto& [r, c] : lock.seatPositions)
                    expireSeat(show.seats[r][c], currentTime);
            }
            lock.released = true;
            return false;
        }
        // Confirm booking
        if (!shows.count(lock.showId)) return false;
        auto& show = shows[lock.showId];
        for (auto& [r, c] : lock.seatPositions) {
            show.seats[r][c].status = SeatStatus::BOOKED;
            show.seats[r][c].bookedBy = lock.userId;
            show.seats[r][c].lockedBy = "";
            show.seats[r][c].lockExpiry = 0;
        }
        string bookingId = "BK_" + lockId;
        bookings[bookingId] = {bookingId, lock.showId, lock.userId,
                               lock.seatPositions};
        lock.confirmed = true;
        return true;
    }

    bool releaseLock(const string& lockId, long currentTime) {
        if (!locks.count(lockId)) return false;
        auto& lock = locks[lockId];
        if (lock.confirmed || lock.released) return false;
        if (shows.count(lock.showId)) {
            auto& show = shows[lock.showId];
            for (auto& [r, c] : lock.seatPositions) {
                show.seats[r][c].status = SeatStatus::AVAILABLE;
                show.seats[r][c].lockedBy = "";
                show.seats[r][c].lockExpiry = 0;
            }
        }
        lock.released = true;
        return true;
    }

    // Part 3: Sliding window — find N contiguous available seats in earliest row
    vector<pair<int,int>> findContiguousSeats(const string& showId, int n, long currentTime = 0) {
        if (!shows.count(showId)) return {};
        auto& show = shows[showId];
        for (int r = 0; r < show.rows; r++) {
            int count = 0, start = -1;
            for (int c = 0; c < show.cols; c++) {
                expireSeat(show.seats[r][c], currentTime);
                if (show.seats[r][c].status == SeatStatus::AVAILABLE) {
                    if (count == 0) start = c;
                    count++;
                    if (count == n) {
                        vector<pair<int,int>> result;
                        for (int i = start; i < start + n; i++) result.push_back({r, i});
                        return result;
                    }
                } else {
                    count = 0;
                    start = -1;
                }
            }
        }
        return {};
    }
};

// ─── Test Entry Points ─────────────────────────────────────────────────────

BookingSystem system_instance;

void add_theater(const string& theaterId, const string& name,
                 const string& city) {
    system_instance.addTheater(theaterId, name, city);
}

void add_show(const string& showId, const string& theaterId,
              const string& movie, const string& time, int rows, int cols) {
    system_instance.addShow(showId, theaterId, movie, time, rows, cols);
}

vector<string> search_movies(const string& city) {
    return system_instance.searchMovies(city);
}

vector<pair<int,int>> get_available_seats(const string& showId,
                                          long currentTime = 0) {
    return system_instance.getAvailableSeats(showId, currentTime);
}

bool book_seats(const string& bookingId, const string& showId,
                const vector<pair<int,int>>& seatPositions,
                const string& userId, long currentTime = 0) {
    return system_instance.bookSeats(bookingId, showId, seatPositions,
                                     userId, currentTime);
}

string lock_seats(const string& showId,
                  const vector<pair<int,int>>& seatPositions,
                  const string& userId, int ttlMinutes, long currentTime) {
    return system_instance.lockSeats(showId, seatPositions, userId,
                                     ttlMinutes, currentTime);
}

bool confirm_booking(const string& lockId, long currentTime) {
    return system_instance.confirmBooking(lockId, currentTime);
}

bool release_lock(const string& lockId, long currentTime) {
    return system_instance.releaseLock(lockId, currentTime);
}

// ─── Ops simulator (used by spec-based tests) ──────────────────────────────

#include <memory>

struct ShowOp {
    string kind;
    string s1;
    string s2;
    string s3;
    string s4;
    int    i1;
    int    i2;
    int    i3;
};

static vector<pair<int,int>> parse_seats(const string& s) {
    vector<pair<int,int>> out;
    string cur;
    auto flush = [&](){
        if (cur.empty()) return;
        size_t comma = cur.find(',');
        if (comma == string::npos) { cur.clear(); return; }
        int r = stoi(cur.substr(0, comma));
        int c = stoi(cur.substr(comma+1));
        out.push_back({r, c});
        cur.clear();
    };
    for (char ch : s) {
        if (ch == ';') flush();
        else cur.push_back(ch);
    }
    flush();
    return out;
}

vector<string> show_simulate(vector<ShowOp> ops) {
    vector<string> out;
    unique_ptr<BookingSystem> sys(new BookingSystem());
    vector<string> lockSlots(32, "");
    vector<pair<int,int>> last_contig;
    for (const auto& op : ops) {
        const string& k = op.kind;
        if (k == "new") {
            sys.reset(new BookingSystem());
            for (auto& s : lockSlots) s.clear();
            last_contig.clear();
            out.push_back("ok");
        } else if (k == "add_theater") {
            sys->addTheater(op.s1, op.s2, op.s3); out.push_back("ok");
        } else if (k == "add_show") {
            sys->addShow(op.s1, op.s2, op.s3, op.s4, op.i1, op.i2); out.push_back("ok");
        } else if (k == "movies_count") {
            out.push_back(to_string((int)sys->searchMovies(op.s1).size()));
        } else if (k == "movies_contains") {
            auto m = sys->searchMovies(op.s1);
            out.push_back(find(m.begin(), m.end(), op.s2) != m.end() ? "yes" : "no");
        } else if (k == "available_count") {
            out.push_back(to_string((int)sys->getAvailableSeats(op.s1, (long)op.i1).size()));
        } else if (k == "available_has") {
            auto v = sys->getAvailableSeats(op.s1, (long)op.i1);
            bool found = false;
            for (auto& p : v) if (p.first == op.i2 && p.second == op.i3) { found = true; break; }
            out.push_back(found ? "yes" : "no");
        } else if (k == "book") {
            out.push_back(sys->bookSeats(op.s1, op.s2, parse_seats(op.s3), op.s4, (long)op.i1) ? "ok" : "fail");
        } else if (k == "lock") {
            string lid = sys->lockSeats(op.s1, parse_seats(op.s2), op.s3, op.i1, (long)op.i2);
            if (op.i3 >= 0 && op.i3 < (int)lockSlots.size()) lockSlots[op.i3] = lid;
            out.push_back(lid);
        } else if (k == "lock_at") {
            out.push_back(lockSlots[op.i3]);
        } else if (k == "confirm") {
            out.push_back(sys->confirmBooking(lockSlots[op.i3], (long)op.i1) ? "ok" : "fail");
        } else if (k == "release") {
            out.push_back(sys->releaseLock(lockSlots[op.i3], (long)op.i1) ? "ok" : "fail");
        } else if (k == "release_id") {
            out.push_back(sys->releaseLock(op.s1, (long)op.i1) ? "ok" : "fail");
        } else if (k == "find_contig") {
            last_contig = sys->findContiguousSeats(op.s1, op.i1, (long)op.i2);
            out.push_back(to_string((int)last_contig.size()));
        } else if (k == "contig_at") {
            if (op.i1 < 0 || op.i1 >= (int)last_contig.size()) out.push_back("");
            else out.push_back(to_string(last_contig[op.i1].first) + "," + to_string(last_contig[op.i1].second));
        } else {
            out.push_back("unknown:" + k);
        }
    }
    return out;
}

// ─── Main ──────────────────────────────────────────────────────────────────

#ifndef RUNNING_TESTS
int main() {
    add_theater("T1", "PVR Phoenix", "Mumbai");
    add_show("S1", "T1", "Inception", "18:00", 5, 10);

    auto movies = search_movies("Mumbai");
    cout << "Movies in Mumbai: " << movies.size() << endl;

    auto seats = get_available_seats("S1");
    cout << "Available seats: " << seats.size() << endl;

    bool ok = book_seats("B1", "S1", {{0,0}, {0,1}}, "user1");
    cout << "Booked: " << (ok ? "YES" : "NO") << endl;

    return 0;
}
#endif

File: problems/tier2-intermediate/015-bookmyshow/solution.java
// BookMyShow — Solution (Java)
import java.util.*;

class ShowOp {
    public String kind;
    public String s1;
    public String s2;
    public String s3;
    public String s4;
    public int i1;
    public int i2;
    public int i3;

    public ShowOp(String kind, String s1, String s2, String s3, String s4, int i1, int i2, int i3) {
        this.kind = kind;
        this.s1 = s1;
        this.s2 = s2;
        this.s3 = s3;
        this.s4 = s4;
        this.i1 = i1;
        this.i2 = i2;
        this.i3 = i3;
    }
}

enum SeatStatus { AVAILABLE, LOCKED, BOOKED }

class Seat {
    public int row, col;
    public SeatStatus status;
    public String lockedBy;
    public long lockExpiry;
    public String bookedBy;

    public Seat(int row, int col) {
        this.row = row;
        this.col = col;
        this.status = SeatStatus.AVAILABLE;
        this.lockedBy = "";
        this.lockExpiry = 0;
        this.bookedBy = "";
    }
}

class Show {
    public String id;
    public String theaterId;
    public String movie;
    public String time;
    public int rows;
    public int cols;
    public Seat[][] seats;

    public Show(String id, String theaterId, String movie, String time, int rows, int cols) {
        this.id = id;
        this.theaterId = theaterId;
        this.movie = movie;
        this.time = time;
        this.rows = rows;
        this.cols = cols;
        this.seats = new Seat[rows][cols];
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                this.seats[r][c] = new Seat(r, c);
    }
}

class Theater {
    public String id;
    public String name;
    public String city;
    public Theater(String id, String name, String city) {
        this.id = id; this.name = name; this.city = city;
    }
}

class Booking {
    public String id;
    public String showId;
    public String userId;
    public List<int[]> seatPositions;
    public Booking(String id, String showId, String userId, List<int[]> sp) {
        this.id = id; this.showId = showId; this.userId = userId; this.seatPositions = sp;
    }
}

class SeatLock {
    public String id;
    public String showId;
    public String userId;
    public List<int[]> seatPositions;
    public long expiry;
    public boolean confirmed;
    public boolean released;

    public SeatLock(String id, String showId, String userId, List<int[]> sp, long expiry) {
        this.id = id; this.showId = showId; this.userId = userId;
        this.seatPositions = sp; this.expiry = expiry;
        this.confirmed = false; this.released = false;
    }
}

class BookingSystem {
    Map<String, Theater> theaters = new LinkedHashMap<>();
    Map<String, Show> shows = new LinkedHashMap<>();
    Map<String, Booking> bookings = new LinkedHashMap<>();
    Map<String, SeatLock> locks = new LinkedHashMap<>();
    Map<String, TreeSet<String>> cityMovies = new LinkedHashMap<>();
    int lockCounter = 0;

    private void expireSeat(Seat seat, long currentTime) {
        if (seat.status == SeatStatus.LOCKED && currentTime >= seat.lockExpiry) {
            seat.status = SeatStatus.AVAILABLE;
            seat.lockedBy = "";
            seat.lockExpiry = 0;
        }
    }

    public void addTheater(String theaterId, String name, String city) {
        theaters.put(theaterId, new Theater(theaterId, name, city));
    }

    public void addShow(String showId, String theaterId, String movie, String time, int rows, int cols) {
        Show show = new Show(showId, theaterId, movie, time, rows, cols);
        shows.put(showId, show);
        Theater t = theaters.get(theaterId);
        if (t != null) {
            cityMovies.computeIfAbsent(t.city, k -> new TreeSet<>()).add(movie);
        }
    }

    public List<String> searchMovies(String city) {
        TreeSet<String> set = cityMovies.get(city);
        if (set == null) return new ArrayList<>();
        return new ArrayList<>(set);
    }

    public List<int[]> getAvailableSeats(String showId, long currentTime) {
        Show show = shows.get(showId);
        if (show == null) return new ArrayList<>();
        List<int[]> result = new ArrayList<>();
        for (int r = 0; r < show.rows; r++) {
            for (int c = 0; c < show.cols; c++) {
                expireSeat(show.seats[r][c], currentTime);
                if (show.seats[r][c].status == SeatStatus.AVAILABLE)
                    result.add(new int[]{r, c});
            }
        }
        return result;
    }

    public boolean bookSeats(String bookingId, String showId, List<int[]> seatPositions,
                              String userId, long currentTime) {
        Show show = shows.get(showId);
        if (show == null) return false;
        for (int[] pos : seatPositions) {
            int r = pos[0], c = pos[1];
            if (r < 0 || r >= show.rows || c < 0 || c >= show.cols) return false;
            expireSeat(show.seats[r][c], currentTime);
            if (show.seats[r][c].status != SeatStatus.AVAILABLE) return false;
        }
        for (int[] pos : seatPositions) {
            int r = pos[0], c = pos[1];
            show.seats[r][c].status = SeatStatus.BOOKED;
            show.seats[r][c].bookedBy = userId;
        }
        bookings.put(bookingId, new Booking(bookingId, showId, userId, seatPositions));
        return true;
    }

    public String lockSeats(String showId, List<int[]> seatPositions, String userId,
                             int ttlMinutes, long currentTime) {
        Show show = shows.get(showId);
        if (show == null) return "";
        for (int[] pos : seatPositions) {
            int r = pos[0], c = pos[1];
            if (r < 0 || r >= show.rows || c < 0 || c >= show.cols) return "";
            expireSeat(show.seats[r][c], currentTime);
            if (show.seats[r][c].status != SeatStatus.AVAILABLE) return "";
        }
        String lockId = "LOCK_" + (++lockCounter);
        long expiry = currentTime + (long) ttlMinutes * 60;
        for (int[] pos : seatPositions) {
            int r = pos[0], c = pos[1];
            show.seats[r][c].status = SeatStatus.LOCKED;
            show.seats[r][c].lockedBy = userId;
            show.seats[r][c].lockExpiry = expiry;
        }
        locks.put(lockId, new SeatLock(lockId, showId, userId, seatPositions, expiry));
        return lockId;
    }

    public boolean confirmBooking(String lockId, long currentTime) {
        SeatLock lock = locks.get(lockId);
        if (lock == null) return false;
        if (lock.confirmed || lock.released) return false;
        if (currentTime >= lock.expiry) {
            Show show = shows.get(lock.showId);
            if (show != null) {
                for (int[] pos : lock.seatPositions) expireSeat(show.seats[pos[0]][pos[1]], currentTime);
            }
            lock.released = true;
            return false;
        }
        Show show = shows.get(lock.showId);
        if (show == null) return false;
        for (int[] pos : lock.seatPositions) {
            int r = pos[0], c = pos[1];
            show.seats[r][c].status = SeatStatus.BOOKED;
            show.seats[r][c].bookedBy = lock.userId;
            show.seats[r][c].lockedBy = "";
            show.seats[r][c].lockExpiry = 0;
        }
        String bookingId = "BK_" + lockId;
        bookings.put(bookingId, new Booking(bookingId, lock.showId, lock.userId, lock.seatPositions));
        lock.confirmed = true;
        return true;
    }

    public boolean releaseLock(String lockId, long currentTime) {
        SeatLock lock = locks.get(lockId);
        if (lock == null) return false;
        if (lock.confirmed || lock.released) return false;
        Show show = shows.get(lock.showId);
        if (show != null) {
            for (int[] pos : lock.seatPositions) {
                int r = pos[0], c = pos[1];
                show.seats[r][c].status = SeatStatus.AVAILABLE;
                show.seats[r][c].lockedBy = "";
                show.seats[r][c].lockExpiry = 0;
            }
        }
        lock.released = true;
        return true;
    }

    public List<int[]> findContiguousSeats(String showId, int n, long currentTime) {
        Show show = shows.get(showId);
        if (show == null) return new ArrayList<>();
        for (int r = 0; r < show.rows; r++) {
            int count = 0, start = -1;
            for (int c = 0; c < show.cols; c++) {
                expireSeat(show.seats[r][c], currentTime);
                if (show.seats[r][c].status == SeatStatus.AVAILABLE) {
                    if (count == 0) start = c;
                    count++;
                    if (count == n) {
                        List<int[]> result = new ArrayList<>();
                        for (int i = start; i < start + n; i++) result.add(new int[]{r, i});
                        return result;
                    }
                } else {
                    count = 0;
                    start = -1;
                }
            }
        }
        return new ArrayList<>();
    }
}

public class Solution {
    private static List<int[]> parseSeats(String s) {
        List<int[]> out = new ArrayList<>();
        if (s == null || s.isEmpty()) return out;
        for (String tok : s.split(";")) {
            if (tok.isEmpty()) continue;
            int comma = tok.indexOf(',');
            if (comma < 0) continue;
            int r = Integer.parseInt(tok.substring(0, comma));
            int c = Integer.parseInt(tok.substring(comma + 1));
            out.add(new int[]{r, c});
        }
        return out;
    }

    public static List<String> show_simulate(List<ShowOp> ops) {
        List<String> out = new ArrayList<>();
        BookingSystem sys = new BookingSystem();
        String[] lockSlots = new String[32];
        for (int i = 0; i < 32; i++) lockSlots[i] = "";
        List<int[]> lastContig = new ArrayList<>();

        for (ShowOp op : ops) {
            String k = op.kind;
            if ("new".equals(k)) {
                sys = new BookingSystem();
                for (int i = 0; i < 32; i++) lockSlots[i] = "";
                lastContig = new ArrayList<>();
                out.add("ok");
            } else if ("add_theater".equals(k)) {
                sys.addTheater(op.s1, op.s2, op.s3); out.add("ok");
            } else if ("add_show".equals(k)) {
                sys.addShow(op.s1, op.s2, op.s3, op.s4, op.i1, op.i2); out.add("ok");
            } else if ("movies_count".equals(k)) {
                out.add(Integer.toString(sys.searchMovies(op.s1).size()));
            } else if ("movies_contains".equals(k)) {
                List<String> m = sys.searchMovies(op.s1);
                out.add(m.contains(op.s2) ? "yes" : "no");
            } else if ("available_count".equals(k)) {
                out.add(Integer.toString(sys.getAvailableSeats(op.s1, (long) op.i1).size()));
            } else if ("available_has".equals(k)) {
                List<int[]> v = sys.getAvailableSeats(op.s1, (long) op.i1);
                boolean found = false;
                for (int[] p : v) if (p[0] == op.i2 && p[1] == op.i3) { found = true; break; }
                out.add(found ? "yes" : "no");
            } else if ("book".equals(k)) {
                out.add(sys.bookSeats(op.s1, op.s2, parseSeats(op.s3), op.s4, (long) op.i1) ? "ok" : "fail");
            } else if ("lock".equals(k)) {
                String lid = sys.lockSeats(op.s1, parseSeats(op.s2), op.s3, op.i1, (long) op.i2);
                if (op.i3 >= 0 && op.i3 < lockSlots.length) lockSlots[op.i3] = lid;
                out.add(lid);
            } else if ("lock_at".equals(k)) {
                out.add(lockSlots[op.i3]);
            } else if ("confirm".equals(k)) {
                out.add(sys.confirmBooking(lockSlots[op.i3], (long) op.i1) ? "ok" : "fail");
            } else if ("release".equals(k)) {
                out.add(sys.releaseLock(lockSlots[op.i3], (long) op.i1) ? "ok" : "fail");
            } else if ("release_id".equals(k)) {
                out.add(sys.releaseLock(op.s1, (long) op.i1) ? "ok" : "fail");
            } else if ("find_contig".equals(k)) {
                lastContig = sys.findContiguousSeats(op.s1, op.i1, (long) op.i2);
                out.add(Integer.toString(lastContig.size()));
            } else if ("contig_at".equals(k)) {
                if (op.i1 < 0 || op.i1 >= lastContig.size()) out.add("");
                else out.add(lastContig.get(op.i1)[0] + "," + lastContig.get(op.i1)[1]);
            } else {
                out.add("unknown:" + k);
            }
        }
        return out;
    }
}
